### PR TITLE
WR#317102 add extra setting for Matomo to allow custom piwik.js library

### PR DESCRIPTION
**Pull request by yao9394**

### CODE REVIEW

File: tool/matomo/classes/tool/tool.php [hunk 1]
```diff
@@ -61,6 +62,11 @@ public function form_add_settings_elements(\MoodleQuickForm &$mform) {
         $mform->setType('siteurl', PARAM_TEXT);
         $mform->addRule('siteurl', get_string('required'), 'required', null, 'client');
 
+        $mform->addElement('text', 'piwikurl', get_string('piwikurl', 'watool_matomo'));
+        $mform->addHelpButton('piwikurl', 'piwikurl', 'watool_matomo');
+        $mform->setType('piwikurl', PARAM_TEXT);
```
Comment (brendanheywood) [inline tool/matomo/classes/tool/tool.php]: perhaps PARAM_URL instead?

File: tool/matomo/classes/tool/tool.php [hunk 2]
```diff
@@ -93,6 +99,18 @@ public function form_validate(&$data, &$files, &$errors) {
                 $errors['siteurl'] = get_string('error:siteurltrailingslash', 'watool_matomo');
             }
         }
+
+        if (!empty($data['piwikurl']) && empty(clean_param($data['piwikurl'], PARAM_URL))) {
```
Comment (brendanheywood) [inline tool/matomo/classes/tool/tool.php]: I think you should get that for free using PARAM_URL ?

File: tool/matomo/lang/en/watool_matomo.php [hunk 3]
```diff
@@ -28,6 +28,8 @@
 $string['pluginname'] = 'Matomo Analytics (formerly Piwik)';
 $string['siteurl'] = 'Analytics URL';
 $string['siteurl_help'] = 'Enter your Matomo URL without http(s) or a trailing slash';
+$string['piwikurl'] = 'Piwik.js URL';
+$string['piwikurl_help'] = 'Enter your piwik.js URL without http(s) or a trailing slash, will be the same as Analytics URL if you leave it empty';
```
Comment (brendanheywood) [inline tool/matomo/lang/en/watool_matomo.php]: I would make this much more clear what the intent is, I'll clearly label is with 'alternate' and say that in the help that it is optional and used for when you want to host your own version of the client code

File: tool/matomo/classes/tool/tool.php [hunk 4]
```diff
@@ -41,6 +41,7 @@ public function get_tracking_code() {
         $template = new \stdClass();
         $template->siteid = $settings['siteid'];
         $template->siteurl = $settings['siteurl'];
+        $template->piwikurl = isset($settings['piwikurl'])? $settings['piwikurl']: '';
```
Comment (brendanheywood) [inline tool/matomo/classes/tool/tool.php]: I would change this so it defaults to siteurl and remove all the 'if' login inside the template

File: tool/matomo/templates/tracking_code.mustache [hunk 5]
```diff
@@ -51,10 +51,17 @@
 
     function embedTrackingCode() {
         var u='//{{{siteurl}}}/';
+        var p='//{{{piwikurl}}}/';
+        var piwik = u;
+
+        if (p != '///') {
```
Comment (brendanheywood) [inline tool/matomo/templates/tracking_code.mustache]: Id simplify this and do that on the php side, see

https://github.com/catalyst/moodle-tool_webanalytics/pull/19/files#r311871053

File: tool/matomo/classes/tool/tool.php [hunk 6]
```diff
@@ -41,6 +41,8 @@ public function get_tracking_code() {
         $template = new \stdClass();
         $template->siteid = $settings['siteid'];
         $template->siteurl = $settings['siteurl'];
+        $usepiwik = (isset($settings['piwikurl']) && !empty($settings['piwikurl']));
```
Comment (dmitriim) [inline tool/matomo/classes/tool/tool.php]: I found that 'piwikurl' is a bit confusing in this case. What do you think about renaming it to  'piwikjsurl' or ''custompiwikjsurl''? or something similar

File: tool/matomo/classes/tool/tool.php [hunk 7]
```diff
@@ -41,6 +41,8 @@ public function get_tracking_code() {
         $template = new \stdClass();
         $template->siteid = $settings['siteid'];
         $template->siteurl = $settings['siteurl'];
+        $usepiwik = (isset($settings['piwikjsurl']) && !empty($settings['piwikjsurl']));
```
Comment (dmitriim) [inline tool/matomo/classes/tool/tool.php]: Sorry, John.  But $usepiwik is confusing as well. I should point to it first time. Maybe $custompiwikjs?

**Pull request merged**